hazelcast/src/main/java/com/hazelcast/query/impl/AndResultSet.java

```diff
@@ -161,6 +150,14 @@ public void remove() {
 
     @Override
     public int size() {
-        return setSmallest.size();
+        if (size == -1) {
```

Comment:
You were right. I've identified edge cases where the overhead of the size() method was significant (i.e. slowing down the query 5x times). 
Fixed.